multi_token_support_program/src/main.leo

```diff
@@ -0,0 +1,1202 @@
+import credits.aleo;
+
+// The 'mtsp' program.
+program multi_token_support_program.aleo {
+  record Token {
+    owner: address,
+    amount: u128,
+    token_id: field,
+    external_authorization_required: bool,
+    authorized_until: u32
+  }
+
+  struct TokenMetadata {
+    token_id: field,
+    name: u128, // ASCII text represented in bits, and the u128 value of the bitstring
+    symbol: u128, // ASCII text represented in bits, and the u128 value of the bitstring
+    decimals: u8,
+    supply: u128,
+    max_supply: u128,
+    admin: address,
+    external_authorization_required: bool, // whether or not this token requires authorization from an external program before transferring
+    external_authorization_party: address
+  }
+
+  struct TokenOwner {
+    account: address,
+    token_id: field
+  }
+
+  struct Balance {
+    token_id: field,
+    account: address,
+    balance: u128,
+    authorized_until: u32
+  }
+
+  struct Allowance {
+    account: address,
+    spender: address,
+    token_id: field
+  }
+
+  const CREDITS_RESERVED_TOKEN_ID: field = 3443843282313283355522573239085696902919850365217539366784739393210722344986field;
+  mapping registered_tokens: field => TokenMetadata;
+  mapping balances: field => Balance; // hash(token_id, account) => Balance
+  mapping authorized_balances: field => Balance; // hash(token_id, account) => Balance
+  mapping allowances: field => u128; // hash(token_id, account, spender) => Allowance
+
+  // mapping struct_balances: TokenOwner => Balance;
+  // mapping struct_authorized_balances: TokenOwner => Balance;
+  // mapping struct_allowances: Allowance => u128;
+
+  // -------------------------
+  // Called by token admins
+  // -------------------------
+
+  async transition initialize() -> Future {
+    return finalize_initialize();
+  }
+
+  async function finalize_initialize() {
+    // Initialize the CREDITS_RESERVED_TOKEN_ID token
+    let credits_reserved_token: TokenMetadata = TokenMetadata {
+      token_id: CREDITS_RESERVED_TOKEN_ID,
+      name: 1095517519u128,
+      symbol: 1095517519u128,
+      decimals: 6u8,
+      supply: 1_500_000_000_000_000u128,
+      max_supply: 1_500_000_000_000_000u128,
+      admin: multi_token_support_program.aleo,
+      external_authorization_required: false,
+      external_authorization_party: multi_token_support_program.aleo
+    };
+
+    registered_tokens.set(CREDITS_RESERVED_TOKEN_ID, credits_reserved_token);
+  }
+
+  async transition register_token(
+    token_id: field,
```

Comment:
Maybe it's a style or preference but not very consequential but should we make these all `public` as they get revealed later anyways?